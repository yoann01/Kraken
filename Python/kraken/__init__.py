"""Kraken Framework."""

import logging
import os

__all__ = ['core', 'helpers', 'plugins', 'ui']


krakenPath = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..'))
if os.environ.get('KRAKEN_PATH', None) is None:
    os.environ['KRAKEN_PATH'] = krakenPath

krakenExtsPath = os.path.join(krakenPath, 'Exts')
if 'FABRIC_EXTS_PATH' in os.environ:
    if krakenExtsPath not in os.environ['FABRIC_EXTS_PATH']:
        os.environ['FABRIC_EXTS_PATH'] = krakenExtsPath + os.pathsep + os.environ['FABRIC_EXTS_PATH']

krakenPresetsPath = os.path.join(krakenPath, 'Presets', 'DFG')
if 'FABRIC_DFG_PATH' in os.environ:
    if krakenPresetsPath not in os.environ['FABRIC_DFG_PATH']:
        os.environ['FABRIC_DFG_PATH'] = krakenPresetsPath + os.pathsep + os.environ['FABRIC_DFG_PATH']


logging.basicConfig(format='[KRAKEN] %(levelname)s: %(message)s', level=logging.INFO)

# Custom inform level for use with UI label getting added to the status bar.
logging.INFORM = 25
logging.addLevelName(logging.INFORM, 'INFORM')
logging.Logger.inform = lambda inst, msg, *args, **kwargs: inst.log(logging.INFORM, msg, *args, **kwargs)


def release():
    print "RELEASE THE KRAKEN!!!!"
