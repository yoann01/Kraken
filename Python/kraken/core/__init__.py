"""Kraken Core."""

VERSION_MAJOR = 1
VERSION_MINOR = 2
VERSION_BUILD = 6
VERSION_SUFFIX = ""


def getVersion():
    """Contatenates the version globals and returns the current version of
    Kraken.

    Returns:
        str: Current version of Kraken.

    """

    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString
