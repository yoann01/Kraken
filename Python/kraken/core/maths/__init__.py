"""Kraken - math module."""

from math_object import MathObject
from vec2 import Vec2
from vec3 import Vec3
from vec4 import Vec4
from quat import Quat
from euler import Euler
from xfo import Xfo
from mat33 import Mat33
from mat44 import Mat44
from rotation_order import RotationOrder
from color import Color
from constants import DEG_TO_RAD, RAD_TO_DEG, PI


def Math_radToDeg(val):
    """Converts radians to degrees.

    Args:
        val (float): value to convert to degrees.

    Returns:
        float: Value in degrees.

    """

    return val * RAD_TO_DEG

def Math_degToRad(val):
    """Converts degrees to radians.

    Args:
        val (float): value to convert to radians.

    Returns:
        float: Value in radians.

    """

    return val * DEG_TO_RAD


def decodeValue(jsonData):
    """Returns a constructed math value based on the provided json data.

    Args:
        jsondata (dict): The JSON data to use to decode into a Math value.

    Returns:
        object: The constructed math value

    """

    if type(jsonData) is not dict:
        return jsonData

    if '__mathObjectClass__' not in jsonData:
        raise Exception("Invalid JSON data for constructing value:" + str(jsonData))

    if jsonData['__mathObjectClass__'] == 'Vec2':
        val = Vec2()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Vec3':
        val = Vec3()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Vec4':
        val = Vec4()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Euler':
        val = Euler()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Quat':
        val = Quat()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Xfo':
        val = Xfo()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Mat33':
        val = Mat33()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Mat44':
        val = Mat44()
        val.jsonDecode(jsonData, decodeValue)
    else:
        raise Exception("Unsupported Math type:" + jsonData['__mathObjectClass__'])

    return val
