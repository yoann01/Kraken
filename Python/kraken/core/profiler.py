"""Kraken - core.profiler module.

Classes:
Profiler - Profiler Object.

"""

import time
import operator


class _ProfilerItem(object):

    def __init__(self, label):
        super(_ProfilerItem, self).__init__()

        t = time.time()
        self.label = label
        self.start = t
        self.end = t
        self.children = []


    def addChild(self, item):
        self.children.append(item)


    def endProfiling(self):
        self.end = time.time()


class Profiler(object):
    """Kraken profiler object for debugging performance issues."""

    __instance = None


    def __init__(self):
        super(Profiler, self).__init__()
        self.reset()


    def reset(self):
        """Resets the profiler for generating a new report"""

        self.__roots = []
        self.__stack = []


    def push(self, label):

        """Adds a new child to the profiling tree and activates it.

        Args:
            label (str): The label of the next level of the stack.

        """

        item = _ProfilerItem(label)
        if len(self.__stack) == 0:
            self.__roots.append(item)
        else:
            self.__stack[-1].addChild(item)

        self.__stack.append(item)


    def pop(self):
        """Deactivates the current item in the tree and returns the profiler to
        the parent item"""

        if len(self.__stack) == 0:
            raise Exception("""Unable to close bracket. Pop has been called more """ +
                            """times than push.""")

        self.__stack[-1].endProfiling()
        self.__stack.pop()


    def generateReport(self, listFunctionTotals=False):
        """Returns a report string containing all the data gathered turing
        profiling.

        Args:
            listFunctionTotals (bool): list information relating to the total time spent in each function.

        Returns:
            str: The profiler report.

        """

        if len(self.__stack) != 0:
            raise Exception("""Profiler brackets not closed properly. """ +
                            """Pop must be called for every call to push. Pop """ +
                            """needs to be called another """ +
                            str(len(self.__stack)) + """ times""")

        report = []
        report.append("--callstack--")

        functions = {}

        def reportItem(item, indent):
            duration = item.end - item.start
            report.append(indent + item.label + ' duration: ' + str(duration))
            if item.label not in functions:
                functions[item.label] = duration
            else:
                functions[item.label] += duration

            for childItem in item.children:
                reportItem(childItem, indent + '  ')

        for rootItem in self.__roots:
            reportItem(rootItem, '  ')

        if listFunctionTotals:
            report.append("--functions--")

            def reverse_numeric(x, y):
                return y < x

            sorted_fns = sorted(functions.items(), key=operator.itemgetter(1),
                                cmp=reverse_numeric)

            for fn_tuple in sorted_fns:
                report.append(str(fn_tuple[1]) + ': ' + fn_tuple[0])

        return '\n'.join(report)


    @classmethod
    def getInstance(cls):
        """This class method returns the singleton instance for the Profiler

        Returns:
            object: The singleton profiler instance.

        """

        if cls.__instance is None:
            cls.__instance = Profiler()

        return cls.__instance
