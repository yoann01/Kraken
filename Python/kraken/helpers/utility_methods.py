from kraken.core.maths.math_object import MathObject
from kraken.core.maths.vec3 import Vec3
from kraken.core.maths.xfo import Xfo
from kraken.core.maths.quat import Quat

from kraken.core.maths import decodeValue


def logHierarchy(kObject):
    """Traverses the given Kraken hierarchy and logs the names of all the objects.

    Args:
        Object: The object to start logging on.

    """

    print kObject.getDecoratedPath()
    for i in xrange(kObject.getNumChildren()):
        child = kObject.getChildByIndex(i)
        logHierarchy(child)


def __convertFromJSON(jsonData):

    if type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertFromJSON(item))
        return newList
    elif type(jsonData) is dict:
        if '__mathObjectClass__' in jsonData.keys():
            return decodeValue(jsonData)
        for key, value in jsonData.iteritems():
            jsonData[key] = __convertFromJSON(value)
    return jsonData


def prepareToLoad(jsonData):
    """Prepares the json data for loading into kraken.

    Args:
        jsonData (dict): The JSON data to be prepared.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __convertFromJSON(jsonData)


def __convertToJSON(jsonData):

    if isinstance(jsonData, MathObject):
        return jsonData.jsonEncode()
    elif type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertToJSON(item))
        return newList
    elif type(jsonData) is dict:
        newDict = {}
        for key, value in jsonData.iteritems():
            newDict[key] = __convertToJSON(value)
        return newDict
    return jsonData


def prepareToSave(jsonData):
    """Prepares the json data for serialization.

    Args:
        jsonData (dict): The JSON data to be prepared.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __convertToJSON(jsonData)


def __mirrorData(jsonData, plane):

    if isinstance(jsonData, Vec3):
        newVec3 = Vec3(jsonData)
        if plane == 0:
            newVec3.x = -newVec3.x
        elif plane == 1:
            newVec3.y = -newVec3.y
        elif plane == 2:
            newVec3.z = -newVec3.z
        return newVec3

    if isinstance(jsonData, Quat):
        newQuat = Quat(jsonData)
        newQuat.mirror(plane)
        return newQuat

    elif isinstance(jsonData, Xfo):
        newXfo = Xfo(jsonData)
        if plane == 0:
            newXfo.tr.x = -newXfo.tr.x
        elif plane == 1:
            newXfo.tr.y = -newXfo.tr.y
        elif plane == 2:
            newXfo.tr.z = -newXfo.tr.z

        newXfo.ori.mirror(plane)
        return newXfo

    elif type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__mirrorData(item, plane))
        return newList

    elif type(jsonData) is dict:
        newDict = {}
        for key, value in jsonData.iteritems():
            newDict[key] = __mirrorData(value, plane)
        return newDict

    return jsonData


def mirrorData(jsonData, plane):
    """Prepares the json data for serialization.

    Args:
        jsonData (dict): The JSON data to be prepared.
        plane (int): The plane to mirror across.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __mirrorData(jsonData, plane)
