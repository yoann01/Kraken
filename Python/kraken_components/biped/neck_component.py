from kraken.core.maths import Vec3, Xfo, Quat

from kraken.core.objects.components.base_example_component import BaseExampleComponent

from kraken.core.objects.joint import Joint
from kraken.core.objects.control import Control
from kraken.core.objects.component_group import ComponentGroup

from kraken.core.objects.operators.kl_operator import KLOperator

from kraken.core.profiler import Profiler


class NeckComponent(BaseExampleComponent):
    """Neck Component"""

    def __init__(self, name="neckBase", parent=None, *args, **kwargs):
        super(NeckComponent, self).__init__(name, parent, *args, **kwargs)

        # ===========
        # Declare IO
        # ===========
        # Declare Inputs Xfos
        self.globalSRTInputTgt = self.createInput(
            'globalSRT',
            dataType='Xfo',
            parent=self.inputHrcGrp).getTarget()

        self.neckBaseInputTgt = self.createInput(
            'neckBase', dataType='Xfo', parent=self.inputHrcGrp).getTarget()

        # Declare Output Xfos
        self.neck01OutputTgt = self.createOutput(
            'neck01', dataType='Xfo', parent=self.outputHrcGrp).getTarget()

        self.neck02OutputTgt = self.createOutput(
            'neck02', dataType='Xfo', parent=self.outputHrcGrp).getTarget()

        self.neckEndOutputTgt = self.createOutput(
            'neckEnd', dataType='Xfo', parent=self.outputHrcGrp).getTarget()

        # Declare Input Attrs
        self.drawDebugInputAttr = self.createInput(
            'drawDebug', dataType='Boolean', value=False, parent=self.cmpInputAttrGrp).getTarget()

        self.rigScaleInputAttr = self.createInput(
            'rigScale', dataType='Float', value=1.0, parent=self.cmpInputAttrGrp).getTarget()

        # Declare Output Attrs


class NeckComponentGuide(NeckComponent):
    """Neck Component Guide"""

    def __init__(self, name='neck', parent=None, *args, **kwargs):

        Profiler.getInstance().push('Construct Neck Component:' + name)
        super(NeckComponentGuide, self).__init__(name, parent, *args, **kwargs)

        # =========
        # Controls
        # =========

        # Guide Controls
        self.neckCtrl = Control('neck', parent=self.ctrlCmpGrp, shape='sphere')
        self.neckCtrl.scalePoints(Vec3(0.5, 0.5, 0.5))
        self.neckMidCtrl = Control('neckMid', parent=self.ctrlCmpGrp, shape='sphere')
        self.neckMidCtrl.scalePoints(Vec3(0.5, 0.5, 0.5))
        self.neckEndCtrl = Control('neckEnd', parent=self.ctrlCmpGrp, shape='sphere')
        self.neckEndCtrl.scalePoints(Vec3(0.5, 0.5, 0.5))

        self.neckCtrlShape = Control('neck', parent=self.ctrlCmpGrp, shape='pin')
        self.neckCtrlShape.rotatePoints(90.0, 0.0, 0.0)
        self.neckCtrlShape.rotatePoints(0.0, 90.0, 0.0)
        self.neckCtrlShape.setColor('orange')
        self.neckMidCtrlShape = Control('neckMid', parent=self.ctrlCmpGrp, shape='pin')
        self.neckMidCtrlShape.rotatePoints(90.0, 0.0, 0.0)
        self.neckMidCtrlShape.rotatePoints(0.0, 90.0, 0.0)
        self.neckMidCtrlShape.setColor('orange')

        # Guide Operator
        self.neckGuideKLOp = KLOperator('guide', 'NeckGuideSolver', 'Kraken')
        self.addOperator(self.neckGuideKLOp)

        # Add Att Inputs
        self.neckGuideKLOp.setInput('drawDebug', self.drawDebugInputAttr)
        self.neckGuideKLOp.setInput('rigScale', self.rigScaleInputAttr)

        # Add Source Inputs
        self.neckGuideKLOp.setInput('sources', [self.neckCtrl, self.neckMidCtrl, self.neckEndCtrl])

        # Add Target Outputs
        self.neckGuideKLOp.setOutput('targets', [self.neckCtrlShape, self.neckMidCtrlShape])


        # Calculate default values
        neckVec = Vec3(0.0, 16.00, -0.75)
        neckMidVec = Vec3(0.0, 16.50, -0.50)
        neckEndVec = Vec3(0.0, 17.00, -0.25)
        upVector = Vec3(0.0, 0.0, -1.0)

        neckOri = Quat()
        neckOri.setFromDirectionAndUpvector((neckMidVec - neckVec).unit(),
                                            ((neckVec + upVector) - neckVec).unit())

        neckMidOri = Quat()
        neckMidOri.setFromDirectionAndUpvector((neckEndVec - neckMidVec).unit(),
                                               ((neckMidVec + upVector) - neckMidVec).unit())

        self.default_data = {
            "name": name,
            "location": "M",
            "neckXfo": Xfo(tr=neckVec, ori=neckOri),
            "neckMidXfo": Xfo(tr=neckMidVec, ori=neckMidOri),
            "neckEndXfo": Xfo(tr=neckEndVec, ori=neckMidOri),
            "neckCrvData": self.neckCtrlShape.getCurveData(),
            "neckMidCrvData": self.neckMidCtrlShape.getCurveData()
        }

        self.loadData(self.default_data)

        Profiler.getInstance().pop()


    # =============
    # Data Methods
    # =============
    def saveData(self):
        """Save the data for the component to be persisted.

        Return:
        The JSON data object

        """

        data = super(NeckComponentGuide, self).saveData()

        data['neckXfo'] = self.neckCtrl.xfo
        data['neckMidXfo'] = self.neckMidCtrl.xfo
        data['neckEndXfo'] = self.neckEndCtrl.xfo

        data['neckCrvData'] = self.neckCtrlShape.getCurveData()
        data['neckMidCrvData'] = self.neckMidCtrlShape.getCurveData()

        return data


    def loadData(self, data):
        """Load a saved guide representation from persisted data.

        Arguments:
            data (object): The JSON data object.

        Returns:
            bool: True if successful.

        """

        super(NeckComponentGuide, self).loadData(data)

        self.neckCtrl.xfo = data.get('neckXfo')
        self.neckMidCtrl.xfo = data.get('neckMidXfo')
        self.neckEndCtrl.xfo = data.get('neckEndXfo')

        self.neckCtrlShape.setCurveData(data.get('neckCrvData'))
        self.neckMidCtrlShape.setCurveData(data.get('neckMidCrvData'))

        # Evaluate guide operators
        self.neckGuideKLOp.evaluate()

        return True


    def getRigBuildData(self):
        """Returns the Guide data used by the Rig Component to define the layout
        of the final rig.

        Return:
        The JSON rig data object.

        """

        data = super(NeckComponentGuide, self).getRigBuildData()

        neckEndXfo = Xfo(tr=self.neckEndCtrl.xfo.tr,
                         ori=self.neckMidCtrlShape.xfo.ori)

        data['neckXfo'] = self.neckCtrlShape.xfo
        data['neckCrvData'] = self.neckCtrlShape.getCurveData()
        data['neckMidXfo'] = self.neckMidCtrlShape.xfo
        data['neckMidCrvData'] = self.neckMidCtrlShape.getCurveData()
        data['neckEndXfo'] = neckEndXfo

        return data


    # ==============
    # Class Methods
    # ==============
    @classmethod
    def getComponentType(cls):
        """Enables introspection of the class prior to construction to determine
        if it is a guide component.

        Returns:
            bool: Whether the component is a guide component.

        """

        return 'Guide'


    @classmethod
    def getRigComponentClass(cls):
        """Returns the corresponding rig component class for this guide
        component class.

        Returns:
            class: The rig component class.

        """

        return NeckComponentRig


class NeckComponentRig(NeckComponent):
    """Neck Component"""

    def __init__(self, name="neck", parent=None):

        Profiler.getInstance().push("Construct Neck Rig Component:" + name)
        super(NeckComponentRig, self).__init__(name, parent)


        # =========
        # Controls
        # =========
        # Neck
        self.neck01Ctrl = Control('neck01', parent=self.ctrlCmpGrp, shape="pin")
        self.neck01Ctrl.setColor("orange")
        self.neck01Ctrl.lockTranslation(True, True, True)
        self.neck01Ctrl.lockScale(True, True, True)

        self.neck01CtrlSpace = self.neck01Ctrl.insertCtrlSpace(name='neck01')

        self.neck02Ctrl = Control('neck02', parent=self.neck01Ctrl, shape="pin")
        self.neck02Ctrl.setColor("orange")
        self.neck02Ctrl.lockTranslation(True, True, True)
        self.neck02Ctrl.lockScale(True, True, True)

        self.neck02CtrlSpace = self.neck02Ctrl.insertCtrlSpace(name='neck02')


        # ==========
        # Deformers
        # ==========
        deformersLayer = self.getOrCreateLayer('deformers')
        self.defCmpGrp = ComponentGroup(self.getName(), self, parent=deformersLayer)
        self.addItem('defCmpGrp', self.defCmpGrp)

        self.neck01Def = Joint('neck01', parent=self.defCmpGrp)
        self.neck01Def.setComponent(self)

        self.neck02Def = Joint('neck02', parent=self.defCmpGrp)
        self.neck02Def.setComponent(self)


        # ==============
        # Constrain I/O
        # ==============
        # Constraint inputs
        neckInputConstraintName = '_'.join([self.neck01CtrlSpace.getName(),
                                            'To',
                                            self.neckBaseInputTgt.getName()])

        self.neckInputCnstr = self.neck01CtrlSpace.constrainTo(
            self.neckBaseInputTgt,
            'Pose',
            maintainOffset=True,
            name=neckInputConstraintName)


        # Constraint outputs
        neck01OutCnstrName = '_'.join([self.neck01OutputTgt.getName(),
                                       'To',
                                       self.neck01Ctrl.getName()])

        self.neck01OutCnstr = self.neck01OutputTgt.constrainTo(
            self.neck01Ctrl,
            'Pose',
            maintainOffset=False,
            name=neck01OutCnstrName)

        neck02OutCnstrName = '_'.join([self.neck02OutputTgt.getName(),
                                       'To',
                                       self.neck02Ctrl.getName()])

        self.neck02OutCnstr = self.neck02OutputTgt.constrainTo(
            self.neck02Ctrl,
            'Pose',
            maintainOffset=False,
            name=neck02OutCnstrName)

        neckEndCnstrName = '_'.join([self.neckEndOutputTgt.getName(),
                                     'To',
                                     self.neck02Ctrl.getName()])

        self.neckEndCnstr = self.neckEndOutputTgt.constrainTo(
            self.neck02Ctrl,
            'Pose',
            maintainOffset=True,
            name=neckEndCnstrName)


        # ==============
        # Add Operators
        # ==============
        # Add Deformer KL Op
        self.neckDeformerKLOp = KLOperator('defConstraint',
                                           'MultiPoseConstraintSolver',
                                           'Kraken')

        self.addOperator(self.neckDeformerKLOp)

        # Add Att Inputs
        self.neckDeformerKLOp.setInput('drawDebug', self.drawDebugInputAttr)
        self.neckDeformerKLOp.setInput('rigScale', self.rigScaleInputAttr)

        # Add Xfo Inputstrl)
        self.neckDeformerKLOp.setInput('constrainers',
                                       [self.neck01Ctrl, self.neck02Ctrl])

        # Add Xfo Outputs
        self.neckDeformerKLOp.setOutput('constrainees',
                                        [self.neck01Def, self.neck02Def])

        Profiler.getInstance().pop()


    def loadData(self, data=None):
        """Load a saved guide representation from persisted data.

        Arguments:
        data -- object, The JSON data object.

        Return:
        True if successful.

        """

        super(NeckComponentRig, self).loadData(data)
        neckXfo = data.get('neckXfo')
        neckCrvData = data.get('neckCrvData')
        neckMidXfo = data.get('neckMidXfo')
        neckMidCrvData = data.get('neckMidCrvData')
        neckEndXfo = data.get('neckEndXfo')

        self.neck01CtrlSpace.xfo = neckXfo
        self.neck01Ctrl.xfo = neckXfo
        self.neck01Ctrl.setCurveData(neckCrvData)

        self.neck02CtrlSpace.xfo = neckMidXfo
        self.neck02Ctrl.xfo = neckMidXfo
        self.neck02Ctrl.setCurveData(neckMidCrvData)


        # ============
        # Set IO Xfos
        # ============
        self.neckBaseInputTgt.xfo = neckXfo
        self.neck01OutputTgt.xfo = neckXfo
        self.neck02OutputTgt.xfo = neckMidXfo
        self.neckEndOutputTgt.xfo = neckEndXfo

        # Evaluate Constraints
        self.neckInputCnstr.evaluate()
        self.neck01OutCnstr.evaluate()
        self.neck02OutCnstr.evaluate()
        self.neckEndCnstr.evaluate()

        # Evaluate Operators
        self.neckDeformerKLOp.evaluate()


from kraken.core.kraken_system import KrakenSystem
ks = KrakenSystem.getInstance()
ks.registerComponent(NeckComponentGuide)
ks.registerComponent(NeckComponentRig)
