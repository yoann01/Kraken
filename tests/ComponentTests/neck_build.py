from kraken import plugins

from kraken_components.biped.neck_component import NeckComponentGuide, NeckComponentRig

from kraken.core.profiler import Profiler
from kraken.helpers.utility_methods import logHierarchy


Profiler.getInstance().push("neck_build")

neckGuide = NeckComponentGuide("neck")
neckGuide.loadData(neckGuide.default_data)


# Save the hand guide data for persistence.
saveData = neckGuide.saveData()

neckGuideData = neckGuide.getRigBuildData()

neck = NeckComponentRig()
neck.loadData(neckGuideData)

builder = plugins.getBuilder()
builder.build(neck)

Profiler.getInstance().pop()


if __name__ == "__main__":
    print Profiler.getInstance().generateReport()
else:
    for each in neck.getItems().values():
        # Only log hierarchy for Layer objects as Layers in this test are added to
        # the component since there is no rig object.
        if each.isTypeOf('Layer'):
            logHierarchy(each)
