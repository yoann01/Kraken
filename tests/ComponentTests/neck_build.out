Failed to find DCC builder. Falling back to Python builder.
building: controls.Neck as: Neck_L_cmp type: ComponentGroup
buildGroup: controls.Neck as: Neck_L_cmp
building: controls.Neck.inputs as: Neck_L_inputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.Neck.inputs as: Neck_L_inputs_hrc
building: controls.Neck.inputs.neckBase as: Neck_L_neckBase_cmpIn type: ComponentInput
buildLocator: controls.Neck.inputs.neckBase as: Neck_L_neckBase_cmpIn
building: controls.Neck.outputs as: Neck_L_outputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.Neck.outputs as: Neck_L_outputs_hrc
building: controls.Neck.outputs.neck as: Neck_L_neck_cmpOut type: ComponentOutput
buildLocator: controls.Neck.outputs.neck as: Neck_L_neck_cmpOut
building: controls.Neck.outputs.neckEnd as: Neck_L_neckEnd_cmpOut type: ComponentOutput
buildLocator: controls.Neck.outputs.neckEnd as: Neck_L_neckEnd_cmpOut
building: controls.Neck.neck as: Neck_L_neck_ctrlSpace type: CtrlSpace
buildGroup: controls.Neck.neck as: Neck_L_neck_ctrlSpace
building: controls.Neck.neck.neck as: Neck_L_neck_ctrl type: Control
buildControl:controls.Neck.neck.neck as: Neck_L_neck_ctrl
building: deformers as: deformers type: Layer
buildLayer: deformers as: deformers
building: deformers.Neck as: Neck_L_cmp type: ComponentGroup
buildGroup: deformers.Neck as: Neck_L_cmp
building: deformers.Neck.neck as: Neck_L_neck_def type: Joint
buildJoint: deformers.Neck.neck as: Neck_L_neck_def
building: controls as: controls type: Layer
buildLayer: controls as: controls
building: controls.Neck as: Neck_L_cmp type: ComponentGroup
buildGroup: controls.Neck as: Neck_L_cmp
building: controls.Neck.inputs as: Neck_L_inputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.Neck.inputs as: Neck_L_inputs_hrc
building: controls.Neck.inputs.neckBase as: Neck_L_neckBase_cmpIn type: ComponentInput
buildLocator: controls.Neck.inputs.neckBase as: Neck_L_neckBase_cmpIn
building: controls.Neck.outputs as: Neck_L_outputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.Neck.outputs as: Neck_L_outputs_hrc
building: controls.Neck.outputs.neck as: Neck_L_neck_cmpOut type: ComponentOutput
buildLocator: controls.Neck.outputs.neck as: Neck_L_neck_cmpOut
building: controls.Neck.outputs.neckEnd as: Neck_L_neckEnd_cmpOut type: ComponentOutput
buildLocator: controls.Neck.outputs.neckEnd as: Neck_L_neckEnd_cmpOut
building: controls.Neck.neck as: Neck_L_neck_ctrlSpace type: CtrlSpace
buildGroup: controls.Neck.neck as: Neck_L_neck_ctrlSpace
building: controls.Neck.neck.neck as: Neck_L_neck_ctrl type: Control
buildControl:controls.Neck.neck.neck as: Neck_L_neck_ctrl
building: deformers.Neck as: Neck_L_cmp type: ComponentGroup
buildGroup: deformers.Neck as: Neck_L_cmp
building: deformers.Neck.neck as: Neck_L_neck_def type: Joint
buildJoint: deformers.Neck.neck as: Neck_L_neck_def
deformers
deformers.Neck:L
deformers.Neck:L.neck
controls
controls.Neck:L
controls.Neck:L.inputs
controls.Neck:L.inputs.neckBase
controls.Neck:L.outputs
controls.Neck:L.outputs.neck
controls.Neck:L.outputs.neckEnd
controls.Neck:L.neck
controls.Neck:L.neck.neck
